### ISSUE TITLE
Jupyter Notebooks don't support Halo

### ISSUE THREAD

**Issue by HBadertscher** [ManrajGrover/halo#32]
While Halo works perfectly fine in the terminal, I can't get Halo to work inside a Jupyter notebook. I am running the following code in the first cell of the notebook:

```
from halo import Halo
from time import sleep
with Halo(text='Loading', spinner='dots'):
    sleep(5)
```
The program runs through without any errors, but no spinner is shown at all.

I figured out, that if you write something to `stdout` with `sys.stdout` prior to calling `Halo`, then the spinner is shown, but each spinner and corresponding text are shown next to each other, i.e. clearing the current line doesn't work.

```
from halo import Halo
import sys
from time import sleep
sys.stdout.write('test\n')
with Halo(text='Loading', spinner='dots'):
    sleep(5)
```
returns the output
```
test
⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ Loading⠴ Loading⠦ Loading⠧ Loading⠇ Loading⠏
Loading⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ Loading⠴ Loading⠦ Loading⠧ Loading⠇
Loading⠏ Loading⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ Loading⠴ Loading⠦ Loading⠧ 
Loading⠇ Loading⠏ Loading⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ Loading⠴ Loading⠦ 
Loading⠧ Loading⠇ Loading⠏ Loading⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ Loading⠴ 
Loading⠦ Loading⠧ Loading⠇ Loading⠏ Loading⠋ Loading⠙ Loading⠹ Loading⠸ Loading⠼ 
Loading⠴ Loading⠦ Loading⠧ Loading
```
Is there something wrong on my end, and do you have an idea how to fix that?
It would be really nice if Halo worked with Jupyter too. Thanks for your efforts and for sharing this library!

**Comment by ManrajGrover** [comment]
@HBadertscher This is one feature I would love to include. I feel we would need to create a widget for jupyter notebook and ipython so that spinner renders as expected. I don't have experience with that, hence will take me a while to finalize the fix and get this working.

If you would like to work on this, feel free to take this one up and I will assign it to you.

**Comment by ManrajGrover** [comment]
Just adding here. If spinner is started in a cell in iPython and not stopped within the same cell, the spinner cannot be stopped. Need to explore this a little more. A widget might fix this.

**Comment by JungWinter** [comment]
@ManrajGrover Can I try to implement this feature? I don't have experience with ipython and jupyter widget too, but it's so interesting to work.

## refer
- [tqdm jupyter integration](https://pypi.python.org/pypi/tqdm#ipython-jupyter-integration)
- [tqdm notebook](https://github.com/tqdm/tqdm/blob/master/tqdm/_tqdm_notebook.py)
- [widget user guide](http://ipywidgets.readthedocs.io/en/latest/user_guide.html)

**Comment by ManrajGrover** [comment]
@JungWinter sure! Please go ahead and give it a try.

**Comment by JungWinter** [comment]
@ManrajGrover I made [prototype](https://github.com/JungWinter/halo/blob/feature-notebook/halo/halo_notebook.py) and can test in [notebook](https://github.com/JungWinter/halo/blob/feature-notebook/examples/notebook.ipynb). 
First, `from ipywidgets import HTML` is tried, but problem was happen about color ansi code.
Second, `from ipywidgets.widgets import Output`, It's work well now. I overwrite `__init__`, `clear`, `_render_frame_`, `start`, `stop_and_persist`.
I checked [examples](https://github.com/ManrajGrover/halo/tree/master/examples) work fine, but not yet contain any jupyter test code.
![capture-gif](https://user-images.githubusercontent.com/13811604/34433276-90ce1fe4-ecc1-11e7-9fb3-61863330db91.gif)

**Comment by JungWinter** [comment]
And when I run `pylint`, SyntaxError is occured. Could you help me to handle this error?

```shell
(halo-venv) halo git:feature-notebook ❯❯❯ pylint --errors-only --rcfile=/Users/WinterJ/Github/halo/.pylintrc --output-format=colorized halo
Traceback (most recent call last):
  File "/Users/WinterJ/.pyenv/versions/halo-venv/bin/pylint", line 11, in <module>
    sys.exit(run_pylint())
  File "/Users/WinterJ/.pyenv/versions/3.6.1/envs/halo-venv/lib/python3.6/site-packages/pylint/__init__.py", line 13, in run_pylint
    Run(sys.argv[1:])
  File "/Users/WinterJ/.pyenv/versions/3.6.1/envs/halo-venv/lib/python3.6/site-packages/pylint/lint.py", line 1260, in __init__
    'init-hook')))
  File "/Users/WinterJ/.pyenv/versions/3.6.1/envs/halo-venv/lib/python3.6/site-packages/pylint/lint.py", line 1361, in cb_init_hook
    exec(value) # pylint: disable=exec-used
  File "<string>", line 8
    ZGljdChfX2ZpbGVfXz1hY3RpdmF0ZV90aGlzKSkK'.decode('base64')
                                            ^
SyntaxError: invalid syntax
```

**Comment by ManrajGrover** [comment]
@JungWinter This looks amazing! I'll have a look at the code soon. Regarding lint error, I'll have a look at it as well while going through your code. Could you please check if this works on `ipython` as well?

**Comment by JungWinter** [comment]
@ManrajGrover Thanks for reviewing! I checked `Halo` already works well. Although `HaloNotebook` doesn't work in `ipython shell`. I think it's not a problem because there is the original `halo`. If there are any problems I didn't recognize, please let me know.

**Comment by ManrajGrover** [comment]
@JungWinter I remember the issue with iPython. If a user does not handle keyboard interrupt and the spinner is running, a `KeyboardInterrupt` is thrown but the spinner keeps spinning. I'm creating an issue for this.

**Comment by JungWinter** [comment]
@ManrajGrover Oh, I didn't check that issue. I worked a little to solve that problem. Please check my comment. https://github.com/ManrajGrover/halo/issues/35#issuecomment-354552242

**Comment by ManrajGrover** [comment]
@JungWinter Please send a PR for this. I'll review the code. 😄 Thank you for working on this!

**Comment by ManrajGrover** [comment]
This is fixed in #40 thanks to @JungWinter! This will be made available soon in next release. I need to do few fixes before making a release.
